### PR TITLE
Add piece per chunk multiplier to chunk download.

### PR DESCRIPTION
**Pull request by RNabel**
1. Adds a multiplier determining the maximum number of pieces per chunk to be downloaded concurrently relative to the minimum required,
2. Adds a limiter to the number of chunks to which this new maximum is applied. The limiter controls that the speed up is only applied to the last N chunks.

This PR has resulted in single-chunk download times to go from ~19s to as low as 2.2s (multiplier val: 2.5). The current multiplier results in speedups of about 3-4x to <5s.

Disclaimer: these measurements were made on a Gbit network with direct connection to the internet backbone.

### CODE REVIEW

**Comment by RNabel** [comment]
The tests are currently exploding because of [this check](https://github.com/NebulousLabs/Sia/blob/master/modules/renter/download.go#L319). Should be an easy fix.

**Comment by DavidVorick** [comment]
This is an amazingly simple change for such a big payoff. We talked about adding it to v1.2.0, but the testing timeline I feel is just too short.

It will be in v1.2.1 for sure though, and we don't plan on taking long to get that out.

File: modules/renter/download.go [hunk 1]
```diff
@@ -493,11 +518,26 @@ func (r *Renter) managedScheduleNewChunks(ds *downloadState) {
 			continue
 		}
 
+		// Determine whether the download should be sped up.
+		su := (nextChunk.download.numChunks - nextChunk.index) <= uint64(numFastChunks)
+		var np = nextChunk.download.erasureCode.MinPieces()
```
Comment (DavidVorick) [inline modules/renter/download.go]: variable names need to read like english. There are exceptions for 'obvious context', but I don't think that applies here.

For example, `cst := newConsensusSetTester` - there's an obvious mapping from cst to ConsensusSetTester. But above, `su` doesn't obviously map to anything.

Same with `np`, it's a bit closer because you have the context of 'MinPieces', but I think still difficult to read.

Also, convention would be to do `np := nextChunk...`, instead of declaring the variable with `var`.

File: modules/renter/download.go [hunk 2]
```diff
@@ -493,11 +518,26 @@ func (r *Renter) managedScheduleNewChunks(ds *downloadState) {
...
+			max := nextChunk.download.erasureCode.NumPieces()
+			n := int(math.Ceil(float64(min) * float64(maxActivePPCMultiplier)))
+			if n > max { // Cap the number of pieces downloaded at the total number of pieces.
+				n = max
+			}
+
+			np = n
+		}
+
+		fmt.Printf("Pieces concurrently downloaded: %d\n", np)
```
Comment (DavidVorick) [inline modules/renter/download.go]: I assume this was leftover from debugging :)

File: modules/renter/download.go [hunk 3]
```diff
@@ -493,11 +518,26 @@ func (r *Renter) managedScheduleNewChunks(ds *downloadState) {
...
 		// Add an incomplete chunk entry for every piece of the download.
-		for i := 0; i < nextChunk.download.erasureCode.MinPieces(); i++ {
+		for i := 0; i < np; i++ {
 			ds.incompleteChunks = append(ds.incompleteChunks, nextChunk)
 		}
-		ds.activePieces += nextChunk.download.erasureCode.MinPieces()
+		ds.activePieces += np
```
Comment (DavidVorick) [inline modules/renter/download.go]: I'm pretty amazed that this update was really this easy. Looks like there's at least one corner case remaining though.

**Comment by DavidVorick** [comment]
The test suite we have is showing a corner case where adding the speedup appears to result in a failed download because there are not enough hosts. The download checker looks at the number of pieces it needs to fetch vs. the number of candidate hosts it has, and will fail if it does not have enough candidate hosts.

It needs to realize that it has enough hosts to download the file, just not enough to perform the speed boost.

File: modules/renter/download.go (hunk 1 above)
Comment (RNabel) [inline modules/renter/download.go]: Thank you for taking the time to get me up to speed on golang best practice :)

File: modules/renter/download.go (hunk 2 above)
Comment (RNabel) [inline modules/renter/download.go]: Yep, removed in the newest revision.

File: modules/renter/download.go [hunk 4]
```diff
@@ -549,8 +611,7 @@ func (r *Renter) managedWaitOnDownloadWork(ds *downloadState) {
 	// If the chunk has completed, perform chunk recovery.
 	if len(cd.completedPieces) == cd.download.erasureCode.MinPieces() {
 		err := cd.recoverChunk()
-		ds.activePieces -= len(cd.completedPieces)
-		cd.completedPieces = make(map[uint64][]byte)
```
Comment (RNabel) [inline modules/renter/download.go]: @DavidVorick could removing this line have other negative consequences? This commit is currently relying on the reference eventually being garbage collected.

As far as I gather from the surrounding code, this line was needed because `len(cd.completedPieces)` was deducted from `ds.activePieces` elsewhere. Since all references to `len(cd.completedPieces)` was replaced with `cd.assignedPieceCount` resetting the map doesn't seem necessary any more.

File: modules/renter/download.go [hunk 5]
```diff
@@ -295,13 +323,34 @@ func (r *Renter) addDownloadToChunkQueue(d *download) {
 			continue
 		}
 
+		// Calculate number of pieces to add to the download list.
+		// If the chunk's download speed should be sped up, the piece
+		// count will be raised above the minimum required.
+		// speedupchunk is true if it is the Nth last chunk where N is defined by `numFastChunks`.
```
Comment (DavidVorick) [inline modules/renter/download.go]: Looks like the comment formatting for all the bulk comments is off. If you use vim, you can get it correct easily by highlighting the block comment in visual mode, and then pressing `gq`. Sublime text has an equivalent, and most code-editors also have some sort of equivalent.

File: modules/renter/download.go (hunk 4 above)
Comment (DavidVorick) [inline modules/renter/download.go]: To be honest, this code is not super well organized and I'm not certain whether this will cause problems or not.

The best way to tell would be an integration test that downloads:

+ a multi-chunk file with enough hosts to utilize the extra pieces
+ a multi-chunk file with seemingly enough hosts to utilize the extra pieces, except some of the hosts have gone offline so really you only have exactly enough hosts available
+ a file with more chunks than hosts that can utilize the full paralleism of the downloader

If it can do all these things, then this code is probably not worse than the existing code.

To be honest, the existing download code is pretty sloppy (my own design) and could use some design iteration. Mostly, it doesn't seem right to me that this file would need to be a whole 500 lines of code - we should be able to achieve everything it does in a lot fewer lines than that (I think, anyway).

File: modules/renter/download.go (hunk 5 above)
Comment (RNabel) [inline modules/renter/download.go]: I'm using neither, and it is rendering ok for me. The comments are in-line with the rest of the block and (mostly) less than 80 characters long. Shall I change them to a block comment, or a concatenate them to a single line?

File: modules/renter/download.go (hunk 4 above)
Comment (RNabel) [inline modules/renter/download.go]: Thanks for letting me know. I was thinking about rewriting this file but don't have the time right now. Maybe we could discuss possible changes to this at some point in the coming weeks. In the meantime, I'll add a couple of tests to check these cases.

**Comment by lukechampine** [comment]
The renter code has changed significantly, and we'll be continuing to overhaul it over the coming weeks. I don't think it makes sense to push this PR through.
